staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/update_validations/listmap_single_key/zz_generated.validations.go

```diff
@@ -0,0 +1,103 @@
+// +build !ignore_autogenerated
+
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Code generated by validation-gen. DO NOT EDIT.
+
+package listmap_single_key
+
+import (
+	fmt "fmt"
+
+	validate "k8s.io/apimachinery/pkg/api/validate"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	field "k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+func init() { localSchemeBuilder.Register(RegisterValidations) }
+
+// RegisterValidations adds validation functions to the given scheme.
+// Public to allow building arbitrary schemes.
+func RegisterValidations(scheme *runtime.Scheme) error {
+	scheme.AddValidationFunc((*M1)(nil), func(obj, oldObj interface{}, subresources ...string) field.ErrorList {
+		if len(subresources) == 0 {
+			return Validate_M1(obj.(*M1), oldObj.(*M1), oldObj != nil, nil)
+		}
+		return field.ErrorList{field.InternalError(nil, fmt.Errorf("no validation found for %T, subresources: %v", obj, subresources))}
+	})
+	scheme.AddValidationFunc((*T1)(nil), func(obj, oldObj interface{}, subresources ...string) field.ErrorList {
+		if len(subresources) == 0 {
+			return Validate_T1(obj.(*T1), oldObj.(*T1), oldObj != nil, nil)
+		}
+		return field.ErrorList{field.InternalError(nil, fmt.Errorf("no validation found for %T, subresources: %v", obj, subresources))}
+	})
+	return nil
+}
+
+func Validate_M1(obj *M1, oldObj *M1, isUpdate bool, fldPath *field.Path) (errs field.ErrorList) {
+	// field M1.K1
+	errs = append(errs,
+		func(obj string, fldPath *field.Path) (errs field.ErrorList) {
+			errs = append(errs, validate.FixedResult(fldPath, obj, true, "M1.K1")...)
+			return
+		}(obj.K1, fldPath.Child("k1"))...)
+
+	// field M1.S
+	var oldS *string = nil
+	if oldObj != nil {
+		oldS = &oldObj.S
+	}
+	errs = append(errs,
+		func(obj string, oldObj *string, fldPath *field.Path) (errs field.ErrorList) {
+			if isUpdate {
+				errs = append(errs, validate.FixedResultUpdate(fldPath, &obj, oldObj, true, "T1.M1.S, UpdateOnly")...)
+			}
+			return
+		}(obj.S, oldS, fldPath.Child("s"))...)
+
+	return errs
+}
+
+func Validate_T1(obj *T1, oldObj *T1, isUpdate bool, fldPath *field.Path) (errs field.ErrorList) {
+	// field T1.LM1
+	var oldLM1 []M1 = nil
+	if oldObj != nil {
+		oldLM1 = oldObj.LM1
+	}
+	errs = append(errs,
+		func(obj []M1, oldObj []M1, fldPath *field.Path) (errs field.ErrorList) {
+			for i, val := range obj {
+				var oldVal *M1 = nil
+				if oldObj != nil {
+					for _, o := range oldObj {
```

Comment:
How do we want to handle non-listmap lists?  For example, consider an update validation for strings which assert that any letter can go from lower- to upper-case, but not vice-versa.  Can I say something like:

```
// +eachVal=+weirdCaseRule
L []string
```

It's not a listmap, so it's unclear to me if `oldVal` can possibly mean anything at all?  The rule I described makes sense for an update from `["aaa", "bbb", "ccc"]` to `["Aaa", "Bbb", "Ccc"]` but not to `["ddd", "eee", "fff"]`.

The point I am trying to grok is: for a plain list (non-listmap) in an UPDATE - is the index of the list enough to form and old/new pair for validation?

It looks like we assume a map-key IS sufficient, but list index is unclear

Reply:
That's right.  We don't make any attempt to correlate non-map list items.

We DO allow the entire list to be correlated, however.  For example, say you want to have an append only list.  That could be done with a validation like:

```
// +appendOnlyList
L []string
```

Where appendOnlyList compares the old list against the new list.